From f11a33d1a5d4aa9f179b918b9ab89598f7383f23 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Thu, 24 Mar 2022 10:38:50 +0200
Subject: [PATCH] [SW-69581] gaudi3: jump to __reset address of zephyr preboot

Similar to running PID/MGMT on ARCs, the __reset address
of the binary is available in the first entry of the
preboot.bin, therefore dereference is needed.

Change-Id: I0224443927b4bb806c804b45081c604741ba4e57
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 3eeb7df..edddffe 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -108,8 +108,14 @@
 
 void image_start(u64 loadaddr)
 {
-	u32 lsb = (u32)loadaddr;
-	void (*image_entry)(void) = (void *)lsb;
+	u32 lsb = (u32) loadaddr;
+	void (*image_entry)(void);
+
+	/*
+	 * Jump to __reset of zephyr. After image_entry, control is not
+	 * expected to return here and continue.
+	 */
+	image_entry = (void *) (*(u32 *) lsb);
 
 	image_entry();
 }
